### PR TITLE
Implement hedgehog state machine models for SDELEG, UTXOW, and CHAIN systems

### PR DESCRIPTION
**Pull request by ruhatch**
This is an initial PR to show a proof-of-concept implementation of the state machine testing of various state transition systems. In the end we would like to add tests for DELEG and various parts of the update system, as well as improving test coverage by including invalid chains in the generation.

### CODE REVIEW

File: src/Cardano/Chain/Txp/GenesisUTxO.hs [hunk 1]
```diff
@@ -5,40 +5,30 @@ where
...
-  (GenesisData(..), unGenesisAvvmBalances, unGenesisNonAvvmBalances)
+import Cardano.Chain.Genesis (unGenesisAvvmBalances, unGenesisNonAvvmBalances)
 import qualified Cardano.Chain.Genesis as Genesis
-import Cardano.Chain.Txp.Tx (TxIn(..), TxOut(..))
 import Cardano.Chain.Txp.UTxO (UTxO)
 import qualified Cardano.Chain.Txp.UTxO as UTxO
-import Cardano.Crypto (hash)
 
 
+-- | Create initial 'UTxO' from balances defined in the genesis block
```
Comment (nc6) [inline src/Cardano/Chain/Txp/GenesisUTxO.hs]:
```suggestion
-- | Create initial 'UTxO' from balances defined in the genesis config
```

The term 'genesis block' has caused confusion multiple times before, so I suggest we avoid it entirely

File: test/Test/Cardano/Chain/Block/Model.hs [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+import qualified Ledger.Delegation as Abstract
+
+import Test.Cardano.Chain.Elaboration.Block
+
+
+--------------------------------------------------------------------------------
+-- CHAIN
+--------------------------------------------------------------------------------
+
+prop_commandCHAIN :: Property
```
Comment (nc6) [inline test/Test/Cardano/Chain/Block/Model.hs:45]: Maybe describe what this property is actually testing?

**Review by nc6 — APPROVED** [review]

File: src/Cardano/Chain/ProtocolConstants.hs [hunk 3]
```diff
@@ -13,19 +13,20 @@ import Cardano.Prelude
...
 import Cardano.Chain.Common.BlockCount (BlockCount(..))
 import Cardano.Chain.Slotting.EpochSlots (EpochSlots(..))
+import Cardano.Chain.Slotting.SlotCount (SlotCount(..))
 
 
 -- | Security parameter expressed in number of slots. It uses chain quality
 --   property. It's basically @blkSecurityParam / chainQualityThreshold@.
-kSlotSecurityParam :: BlockCount -> EpochSlots
-kSlotSecurityParam = EpochSlots . (*) 2 . unBlockCount
+kSlotSecurityParam :: BlockCount -> SlotCount
```
Comment (dnadales) [inline src/Cardano/Chain/ProtocolConstants.hs:21]: Uh, didn't we change `SlotCount` to `EpochSlots` ?

File: test/Test/Cardano/Chain/Block/Model.hs (hunk 2 above)
Comment (dnadales) [reply test/Test/Cardano/Chain/Block/Model.hs:45]: Isn't that described in the comment above?

File: test/Test/Cardano/Chain/Txp/Model.hs [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+modelAddresses :: [Abstract.Addr]
+modelAddresses = Abstract.Addr . Abstract.VKey . Abstract.Owner <$> [0 .. 5]
+
+
+--------------------------------------------------------------------------------
+-- Main Test Export
+--------------------------------------------------------------------------------
+
+tests :: IO Bool
+tests = checkSequential $$discover
```
Comment (dnadales) [inline test/Test/Cardano/Chain/Txp/Model.hs:303]: Shouldn't we be able to check in parallel?

File: test/Test/Cardano/Chain/Txp/Model.hs [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+      , crAppliedTx = Concrete.taTx concreteTxAux
+      }
+
+  callbacks :: [Callback SignalUTXOW ConcreteResult StateUTXOW]
+  callbacks =
+    [ Update update
+    , Ensure $ \_ StateUTXOW { lastResult } _ concreteResult -> do
+      annotateShow lastResult
+      isRight lastResult === isRight (crResult concreteResult)
+    , Require post
```
Comment (dnadales) [inline test/Test/Cardano/Chain/Txp/Model.hs]: Require is a pre-condition. Should `post` be renamed to `pre`?

File: test/Test/Cardano/Chain/Txp/Model.hs [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+        }
+
+      result =
+        STS.applySTS @(UTXOW v) (STS.TRC (env, abstractState, txWits'))
+    in StateUTXOW (fromRight abstractState result) result
+
+  post :: StateUTXOW Symbolic -> SignalUTXOW Symbolic -> Bool
+  post StateUTXOW { abstractState } (SignalUTXOW (Abstract.TxWits tx _)) =
+    all (`elem` M.keys m) (Abstract.txins tx)
+      && Abstract.balance (Abstract.txouts tx)
```
Comment (dnadales) [inline test/Test/Cardano/Chain/Txp/Model.hs:240]: Isn't this being checked by the UTxO rule? Why do we need to check this here?

**Review by dnadales — APPROVED** [review]
Thanks for the great work Ru.

Some considerations:
- there are some repeating patterns regarding testing STS's with hedgehog-SM that we should be able to factor out later on.
- I'm a bit worried by the lack of symmetry between the testing of delegation and chain, and UTxO. In the former we do not make use of symbolic/concrete variables, however in the latter we had to parametrize `UTXOW`. I don't know whether we could reconcile the two approaches.
- I'm also worried (that's what I do apparently) by the additional complexity introduced (specially w.r.t. the use of traces). But since we do not have a better alternative at the moment I guess I'll have to live with it.

But all in all it looks good! Thanks again.

File: src/Cardano/Chain/ProtocolConstants.hs (hunk 3 above)
Comment (ruhatch) [reply src/Cardano/Chain/ProtocolConstants.hs:21]: Only where that makes sense. In this case this isn't a number of slots in an epoch, but instead just a number of slots.

File: test/Test/Cardano/Chain/Txp/Model.hs (hunk 4 above)
Comment (ruhatch) [reply test/Test/Cardano/Chain/Txp/Model.hs:303]: There's only one property, so shouldn't make a difference!

File: test/Test/Cardano/Chain/Txp/Model.hs (hunk 5 above)
Comment (ruhatch) [reply test/Test/Cardano/Chain/Txp/Model.hs]: Yep!

File: test/Test/Cardano/Chain/Txp/Model.hs (hunk 6 above)
Comment (ruhatch) [reply test/Test/Cardano/Chain/Txp/Model.hs:240]: This is for shrinking. If you don't have this here then you can shrink to invalid transactions :(

**Review by mgudemann — APPROVED** [review]

File: test/Test/Cardano/Chain/Block/Model.hs (hunk 2 above)
Comment (ruhatch) [reply test/Test/Cardano/Chain/Block/Model.hs:45]: Hehe, I updated this before Damian saw and confused everyone

**Comment by ruhatch** [comment]
Thanks for the reviews!

bors r+

**Pull request merged**